drv/spartan7-loader/src/main.rs

```diff
@@ -0,0 +1,230 @@
+    drv_stm32h7_spi_server_core::declare_spi_core!(
+        sys.clone(),
+        notifications::SPI_IRQ_MASK
+    )
+}
+
+#[cfg(not(feature = "use-spi-core"))]
+pub fn claim_spi(_sys: &sys_api::Sys) -> drv_spi_api::Spi {
+    task_slot!(SPI, spi);
+    drv_spi_api::Spi::from(SPI.get_task_id())
+}
+
+////////////////////////////////////////////////////////////////////////////////
+
+#[derive(Copy, Clone, PartialEq, Count)]
+enum Trace {
+    FpgaInit,
+    FpgaInitFailed(#[count(children)] Spartan7Error),
+    StartFailed(#[count(children)] SeqError),
+    ContinueBitstreamLoad(usize),
+    WaitForDone,
+    Programmed,
+
+    #[count(skip)]
+    None,
+}
+
+#[derive(Copy, Clone, PartialEq, Count)]
+enum SeqError {
+    AuxFlashError(#[count(children)] drv_auxflash_api::AuxFlashError),
+    SpartanError(#[count(children)] Spartan7Error),
+    AuxChecksumMismatch,
+}
+
+impl From<drv_auxflash_api::AuxFlashError> for SeqError {
+    fn from(v: drv_auxflash_api::AuxFlashError) -> Self {
+        SeqError::AuxFlashError(v)
+    }
+}
+
+impl From<Spartan7Error> for SeqError {
+    fn from(v: Spartan7Error) -> Self {
+        SeqError::SpartanError(v)
+    }
+}
+
+counted_ringbuf!(Trace, 128, Trace::None);
+
+task_slot!(SYS, sys);
+task_slot!(AUXFLASH, auxflash);
+
+#[export_name = "main"]
+fn main() -> ! {
+    match init() {
+        // Set up everything nicely, time to start serving incoming messages.
+        Ok(()) => {
+            let mut server = ServerImpl;
+            let mut buffer = [0; idl::INCOMING_SIZE];
+            loop {
+                idol_runtime::dispatch(&mut buffer, &mut server);
+            }
+        }
+
+        // Initializing the sequencer failed.
+        Err(e) => {
+            // Log that something's broken
+            //
+            // The main sequencer task has ownership over the fault pin and is
+            // responsible for leaving it held low until the FPGA boots, so it
+            // will be stuck low here.
+            ringbuf_entry!(Trace::StartFailed(e));
+
+            // All these moments will be lost in time, like tears in rain...
+            // Time to die.
+            loop {
+                // Sleeping with all bits in the notification mask clear means
+                // we should never be notified --- and if one never wakes up,
+                // the difference between sleeping and dying seems kind of
+                // irrelevant. But, `rustc` doesn't realize that this should
+                // never return, we'll stick it in a `loop` anyway so the main
+                // function can return `!`
+                sys_recv_notification(0);
+            }
+        }
+    }
+}
+
+fn init() -> Result<(), SeqError> {
+    let sys = sys_api::Sys::from(SYS.get_task_id());
+    let dev = claim_spi(&sys).device(drv_spi_api::devices::SPARTAN7_FPGA);
+    let aux = drv_auxflash_api::AuxFlash::from(AUXFLASH.get_task_id());
+
+    init_spartan7_fpga(&sys, &dev, &aux)?;
```

Comment:
hm, i don't totally get why `init_spartan7_fpga` is a separate function, since (AFAIK) it only gets called here, and we don't use it as a separate scope for the `?` operator or similar.  is there any reason for factoring it out into a separate function i'm missing?

Reply:
Historical artifact; fixed in 1572e138